File: 229-majority-element-ii/229-majority-element-ii.py
class Solution:
    def majorityElement(self, nums: List[int]) -> List[int]:
        cand1, cand2, cand1_votes, cand2_votes = 0, 1, 0, 0
        for num in nums:
            if cand1==num:
                cand1_votes+=1
            elif cand2==num:
                cand2_votes+=1
            elif cand1_votes==0:
                cand1=num
                cand1_votes=1
            elif cand2_votes==0:
                cand2=num
                cand2_votes=1
            else:
                cand1_votes-=1
                cand2_votes-=1
            # print(num,cand1,cand2,cand1_votes,cand2_votes)
        res=[]
        # print(cand1,cand2)
        if nums.count(cand1)>len(nums)//3:
            res.append(cand1)
        if nums.count(cand2) >len(nums)//3 and ((not res) or res[-1]!=cand2):
            res.append(cand2)
        return res

File: 229-majority-element-ii/229-majority-element-ii.cpp
class Solution {
public:
    vector<int> majorityElement(vector<int>& nums) {
        if(nums.size()==1) return {nums[0]};
            
        int cand1=0;
        int cand2=1;
        int votes1=0, votes2=0;
        
        for(int i=0;i<nums.size();i++){
            if(nums[i]==cand1) votes1++;
            else if(nums[i]==cand2) votes2++;
            else if(votes1==0) cand1=nums[i], votes1=1;
            else if(votes2==0) cand2=nums[i], votes2=1;
            else votes1--, votes2--;
        }
        // cout<<cand1<<" "<<cand2<<endl;
        vector<int>res;
        
        int count1=0, count2=0;
        for(int i=0;i<nums.size();i++){
            if(nums[i]==cand1) count1++;
            else if(nums[i]==cand2) count2++;
        }
        if(count1>nums.size()/3) res.push_back(cand1);
        if(count2>nums.size()/3) res.push_back(cand2);
        return res;
    }
};